build(cmake): support building with (almost) all system 3rd-party libraries

tearfur opened the pull request:
Superscedes #6900.

Detailed changes made to each third-party dependency:
- libevent
   - Build as subproject.
   - Dropped the deprecated `libevent::event` target in favour of `libevent::core` and `libevent::extra`.
   - Prefer finding the system library using the `find_package` config mode instead of `FindPkgConfig`.
- miniupnpc
   - Prefer finding the system library using the `find_package` config mode instead of `FindPkgConfig`.
   - Use the official naming `miniupnpc::miniupnpc` instead of `miniupnpc::libminiupnpc`.
- libutp, libb64, libpsl: Misc.
- libdeflate
   - Changed the target from `deflate::deflate` to `libdeflate::libdeflate` to closer match the official namespace.
   - Prefer finding the system library using the `find_package` config mode instead of `FindPkgConfig`.
- fast_float
   - Support building with system library.
- fmt
   - Support building with system library.
- rapidjson
   - Support building with system library.
- small
   - Support building with system library.
- utf8cpp
   - Support building with system library.
- WideInteger
   - Support building with system library.
- gtest
   - Support building with system library.
   - Build with submodule gtest by default. (`USE_SYSTEM_GTEST=OFF` by default)
   - Only test system library in Debian CI since only they have that requirement (AFAIK).

Also test building Transmission with using every system library in the CI.

Now wildmat is the only library where we don't support building with system library.

Notes: Added support for building more third-party dependencies installed in the system.

tearfur committed:
build: overhaul tr_add_external_auto_library

tearfur committed:
build: use package-provided CMake config file libevent

tearfur committed:
build: use package-provided CMake config file miniupnpc

tearfur committed:
build: update libutp find module

tearfur committed:
build: use package-provided CMake config file libdeflate

tearfur committed:
build: make LIBNAME an optional parameter

tearfur committed:
build: update libb64 find module

tearfur committed:
build: update libnatpmp find module

tearfur committed:
build: update libpsl find module

tearfur committed:
build: support system fast_float library

tearfur committed:
chore: reformat long brew commands

tearfur committed:
build: support system fmt library

tearfur committed:
build: support system rapidjson library

tearfur committed:
build: support system small library

tearfur committed:
build: support system library utf8cpp

tearfur committed:
build: support system library WideInteger

tearfur committed:
build: support system library gtest

tearfur commented:
The compiler warnings works in mysterious ways... For example, on 4ef3fd7ce34365a67f0aa7db27818aa6d5b57dda, `bitfield.cc` generates these warnings on Ubuntu 24.04 clang  that are not in `main`:

```
[35/249] Building CXX object libtransmission/CMakeFiles/transmission.dir/bitfield.cc.o
In file included from /home/runner/work/transmission/transmission/src/libtransmission/bitfield.cc:11:
/home/runner/work/transmission/transmission/src/libtransmission/../libtransmission/bitfield.h:112:46: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
  112 |         return static_cast<float>(count()) / size();
      |                                            ~ ^~~~~~
/home/runner/work/transmission/transmission/src/libtransmission/bitfield.cc:48:28: warning: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char') [-Wimplicit-int-conversion]
   48 |         array[n - 1] = Val << shift;
      |                      ~ ~~~~^~~~~~~~
/home/runner/work/transmission/transmission/src/libtransmission/bitfield.cc:393:36: warning: implicit conversion loses integer precision: 'int' to 'unsigned char' [-Wimplicit-int-conversion]
  393 |     unsigned char last_mask = 0xff << ((~end) & 7U);
      |                   ~~~~~~~~~   ~~~~~^~~~~~~~~~~~~~~~
3 warnings generated.
```

So with the help of `CMAKE_FIND_LIBRARY_SUFFIXES`, I tried to find out what's different with the compile command, but was surprised to find there are ***no*** differences.

```diff
399,400c399,400
<   "directory": "/root/transmission/obj_main",
<   "command": "/usr/bin/clang++ -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DHAVE_COPY_FILE_RANGE -DHAVE_FALLOCATE64 -DHAVE_FLOCK -DHAVE_GETMNTENT -DHAVE_GETTEXT -DHAVE_LIBINTL_H -DHAVE_MKDTEMP -DHAVE_NGETTEXT -DHAVE_POSIX_FADVISE -DHAVE_POSIX_FALLOCATE -DHAVE_PREAD -DHAVE_PWRITE -DHAVE_SENDFILE64 -DHAVE_SO_REUSEPORT=1 -DHAVE_STATVFS -DHAVE_SYS_STATVFS_H -DPACKAGE_DATA_DIR=\\\"/root/transmission/pfx/share\\\" -DPOSIX -DRAPIDJSON_HAS_STDSTRING=1 -DSMALL_DISABLE_EXCEPTIONS=1 -DUSE_SYSTEM_B64 -DWIDE_INTEGER_DISABLE_FLOAT_INTEROP -DWIDE_INTEGER_DISABLE_IOSTREAM -DWIDE_INTEGER_HAS_LIMB_TYPE_UINT64 -DWITH_INOTIFY -DWITH_OPENSSL -DWITH_UTP -D__TRANSMISSION__ -I/root/transmission/libtransmission/.. -I/root/transmission/obj_main/libtransmission/.. -I/root/transmission/third-party/libutp/include -I/root/transmission/third-party/wildmat -isystem /root/transmission/third-party/fast_float/include -isystem /root/transmission/obj_main/third-party/dht.bld/pfx/include -isystem /root/transmission/third-party/rapidjson/include -isystem /root/transmission/third-party/utfcpp/source -isystem /root/transmission/third-party/wide-integer -isystem /root/transmission/third-party/fmt/include -isystem /root/transmission/third-party/small/include -g -std=gnu++17 -W -Wall -Wextra -Wcast-align -Wextra-semi -Wextra-semi-stmt -Wextra-tokens -Wfloat-equal -Wgnu -Winit-self -Wint-in-bool-context -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wreorder-ctor -Wreturn-std-move -Wself-assign -Wself-move -Wsemicolon-before-method-body -Wsentinel -Wshadow -Wsign-compare -Wsometimes-uninitialized -Wstring-conversion -Wsuggest-destructor-override -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -o libtransmission/CMakeFiles/transmission.dir/bitfield.cc.o -c /root/transmission/libtransmission/bitfield.cc",
---
>   "directory": "/root/transmission/obj_branch",
>   "command": "/usr/bin/clang++ -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DHAVE_COPY_FILE_RANGE -DHAVE_FALLOCATE64 -DHAVE_FLOCK -DHAVE_GETMNTENT -DHAVE_GETTEXT -DHAVE_LIBINTL_H -DHAVE_MKDTEMP -DHAVE_NGETTEXT -DHAVE_POSIX_FADVISE -DHAVE_POSIX_FALLOCATE -DHAVE_PREAD -DHAVE_PWRITE -DHAVE_SENDFILE64 -DHAVE_SO_REUSEPORT=1 -DHAVE_STATVFS -DHAVE_SYS_STATVFS_H -DPACKAGE_DATA_DIR=\\\"/root/transmission/pfx/share\\\" -DPOSIX -DRAPIDJSON_HAS_STDSTRING=1 -DSMALL_DISABLE_EXCEPTIONS=1 -DUSE_SYSTEM_B64 -DWIDE_INTEGER_DISABLE_FLOAT_INTEROP -DWIDE_INTEGER_DISABLE_IOSTREAM -DWIDE_INTEGER_HAS_LIMB_TYPE_UINT64 -DWITH_INOTIFY -DWITH_OPENSSL -DWITH_UTP -D__TRANSMISSION__ -I/root/transmission/libtransmission/.. -I/root/transmission/obj_branch/libtransmission/.. -I/root/transmission/third-party/libutp/include -I/root/transmission/third-party/wildmat -isystem /root/transmission/third-party/fast_float/include -isystem /root/transmission/obj_branch/third-party/dht.bld/pfx/include -isystem /root/transmission/third-party/rapidjson/include -isystem /root/transmission/third-party/utfcpp/source -isystem /root/transmission/third-party/wide-integer -isystem /root/transmission/third-party/fmt/include -isystem /root/transmission/third-party/small/include -g -std=gnu++17 -W -Wall -Wextra -Wcast-align -Wextra-semi -Wextra-semi-stmt -Wextra-tokens -Wfloat-equal -Wgnu -Winit-self -Wint-in-bool-context -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wreorder-ctor -Wreturn-std-move -Wself-assign -Wself-move -Wsemicolon-before-method-body -Wsentinel -Wshadow -Wsign-compare -Wsometimes-uninitialized -Wstring-conversion -Wsuggest-destructor-override -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -o libtransmission/CMakeFiles/transmission.dir/bitfield.cc.o -c /root/transmission/libtransmission/bitfield.cc",
```

nevack commented:
Impressive piece of work.

Dunno if we need to support the same for Xcode project, possibly not, as we need to bundle everything.
At least I need to make sure Xcode and CMake builds are compatible once again.

Need some time to wrap my head around this.

tearfur commented:
> At least I need to make sure Xcode and CMake builds are compatible once again.

You might want to hold on a bit on this. Let me at least get the CI green.

tearfur committed:
fix: incorrectly labeled test suites

tearfur commented:
@niol This should allow folks at Debian to do away with https://salsa.debian.org/debian/transmission/-/blob/debian/experimental/debian/patches/use-system-libs.patch completely.

See if this works for you.

tearfur commented on CMakeLists.txt:
```diff
@@ -25,13 +25,16 @@ endif()
     "${PROJECT_SOURCE_DIR}/cmake"
     "${TR_THIRD_PARTY_SOURCE_DIR}/rpavlik-cmake-modules")
 
+list(APPEND CMAKE_IGNORE_PREFIX_PATH
+    "${PROJECT_SOURCE_DIR}")
```
This didn't work as I had hoped...

https://trtc.mikedld.com/buildConfiguration/Transmission_Sanity/431174?showLog=431171_2000000000_2000000000&logFilter=debug

```
[06:11:24]W:	 [Step 1/5] CMake Error at /home/user/buildAgent/work/1f3bff43dd55cfe9/_build/third-party/rapidjson.bld/src/_RapidJSON-build/RapidJSONConfig.cmake:3 (include):
[06:11:24]W:	 [Step 1/5]   include could not find requested file:
[06:11:24]W:	 [Step 1/5] 
[06:11:24]W:	 [Step 1/5]     /home/user/buildAgent/work/1f3bff43dd55cfe9/_build/third-party/rapidjson.bld/src/_RapidJSON-build/RapidJSON-targets.cmake
[06:11:24]W:	 [Step 1/5] Call Stack (most recent call first):
[06:11:24]W:	 [Step 1/5]   cmake/FindRapidJSON.cmake:1 (find_package)
[06:11:24]W:	 [Step 1/5]   cmake/TrMacros.cmake:154 (find_package)
[06:11:24]W:	 [Step 1/5]   CMakeLists.txt:597 (tr_add_external_auto_library)
[06:11:24]W:	 [Step 1/5] 
[06:11:24]W:	 [Step 1/5] 
[06:11:24]W:	 [Step 1/5] CMake Error at /home/user/buildAgent/work/1f3bff43dd55cfe9/_build/third-party/rapidjson.bld/src/_RapidJSON-build/RapidJSONConfig.cmake:17 (get_target_property):
[06:11:24]W:	 [Step 1/5]   get_target_property() called with non-existent target "RapidJSON".
[06:11:24]W:	 [Step 1/5] Call Stack (most recent call first):
[06:11:24]W:	 [Step 1/5]   cmake/FindRapidJSON.cmake:1 (find_package)
[06:11:24]W:	 [Step 1/5]   cmake/TrMacros.cmake:154 (find_package)
[06:11:24]W:	 [Step 1/5]   CMakeLists.txt:597 (tr_add_external_auto_library)
```

tearfur replied:
6f0be21d0f8fe91c22c4c7718458fcbcb902848d Hopefully this will do it.

tearfur commented:
> Dunno if we need to support the same for Xcode project, possibly not, as we need to bundle everything.
> At least I need to make sure Xcode and CMake builds are compatible once again.

Looking at the CI, seems like the Xcode build isn't expected to work with system libraries in the first place? IIUC no changes are needed for the Xcode project.

nevack commented:
> seems like the Xcode build isn't expected to work with system libraries in the first place?

yep, everything is being built from source and bundled with the app.

> IIUC no changes are needed for the Xcode project.

That's right.

Sorry for the initial confusion.

jggimi commented:
I'd be interested in this -- on OpenBSD we build transmission with a patched third-party/fmt in order to downgrade it to match the fmt in the OS ports/packages system.

tearfur commented:
> I'd be interested in this -- on OpenBSD we build transmission with a patched third-party/fmt in order to downgrade it to match the fmt in the OS ports/packages system.

@jggimi Not sure when/whether this PR will be merged. In the meantime, instead of patching the whole `third-party/fmt`, would it be easier to patch `cmake/FindFmt.cmake` to look for the system library?

https://github.com/transmission/transmission/blob/f7373cb6483bd624c065cdc5a3b53908ee9b1902/cmake/FindFmt.cmake#L3-L5

jggimi commented:
> @jggimi Not sure when/whether this PR will be merged. In the meantime, instead of patching the whole `third-party/fmt`, would it be easier to patch `cmake/FindFmt.cmake` to look for the system library?

From previous testing, it looks like #7554 has already solved the integration issue for the OS.  Without backleveling third-party/fmt, transmission(-qt, -gtk) ran without any apparent issues.

tearfur commented:
I'm a bit confused. Did you mean #7595 instead of #7554?

Anyway, glad to hear whatever issue the OpenBSD folks were having are solved now. Just out of curiosity, could you please point me to the bug ticket or mailing list archive that tracks this issue?

jggimi commented:
Sorry, the commit that seems to have solved the issue was  f7373cb: "fixup! chore: move away from fmt/core.h (#7557) (#7595)"

The issue appeared in integration testing of both betas, and  reversion to fmt 10.2.1 was noted in my call for testing in OpenBSD's ports@ mailing list here: https://marc.info/?l=openbsd-ports&m=173406443719778&w=2

tearfur commented:
I'm wondering if there is a problem with our build system that I'm not aware of. I'm having trouble understanding how changing the includes to `#include <fmt/format.h>` could fix something, especially since our CI pipelines are green.

What exactly is the issue? Where can I look at the test output and test code? Please help me wrap my head around it.

What I'm understanding so far is, the `net/transmission 4.1.0-beta.2` port has a build dependency on `devel/fmt` (currently on 10.2.1):

```Makefile
BUILD_DEPENDS +=	devel/fmt
```

But when building the port, the bundled `third-party/fmt` (11.0.2 on this repo) is the one that's being used (which builds fine on our CI), so the `devel/fmt` build dependency is irrelevant:

https://github.com/jggimi/transmission/blob/71838d6b1d6ff639c2659c6cb6f4cc15cf18b2aa/cmake/FindFmt.cmake#L3-L5

And there's an issue that is not showing up in our CI pipelines, causing BSD ports to patch `third-party/fmt` to 10.2.1.

From a [code search](https://github.com/search?q=repo%3Aopenbsd%2Fports%20%22net%2Ftransmission%22&type=code) of the OpenBSD ports repo, `net/transmission` seems to have no reverse dependencies. So it shouldn't break other packages as well (?)

jggimi commented:
If i recall correctly it wasn't a build issue at all.   It was an issue running with mismatched libraries.  

If be happy to try to recreate the problem for you, if you'd like to see some gdb output.

jggimi commented:
It may be a while.  I began a test build, but googltest is now failing and that's a brand new symptom.   I'm not sure why ninja is aborting.  Yet.  (My test build was a checkout of the 4.1.0-beta.2 tag, then tarring up an an archive with recursive submodules.)

jggimi commented:
The problem appears resolved by the "move away from fmt/core.h" commit, as I'd noted.  I retested again today with the main branch, just to make sure.  But, other library mismatches between system and third-party could possibly arise again.  If they do, I know the workaround. :)  

I will guess that the CI test system for OpenBSD doesn't have the fmt package installed, so the problem doesn't occur in CI.

tearfur commented:
@jggimi I spun up an OpenBSD 7.7 VM to do some testing with vanilla `4.1.0-beta.2`. I know the the problem doesn't exist anymore in latest `main`, but I want to figure out the root cause.

tldr I think the cause of the problem is most likely somewhere within the process of linking object files to the test executable, but I don't know how to investigate further.

Some quick facts:
1. I tested with 2 builds. One without `devel/fmt` installed and one with `devel/fmt` installed. The CMake command is the same for both builds.
2. All test cases pass when building without `devel/fmt` installed.
3. Some test cases fail with segfault when building with `devel/fmt` installed.
4. A `libtransmission-test` executable produced without `devel/fmt` installed continues to pass all tests even if `devel/fmt` is installed afterwards. This rules out the possibility of wonky dynamic linking.
5. Running `ldd libtransmission-test` lists the same libraries for both builds. This further supports the previous point.
6. The preprocessor output when building without `devel/fmt` and with `devel/fmt` is the same. This rules out the possibility of something wonky during the preprocessor or compilation.
7. The compiler commands for producing object files (`.cc` -> `.cc.o`) are the same in both builds. (Commands are obtained by setting `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON`)
8. Adding `-v` to the compiler commands and manually running them lists the same include directories.
    <details>
    <summary>Example command and output</summary>

    ```console
    # /usr/bin/c++ -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DLIBTRANSMISSION_TEST_ASSETS_DIR="/home/test/transmission/tests/libtransmission/assets" -DSMALL_DISABLE_EXCEPTIONS=1 -DWIDE_INTEGER_HAS_LIMB_TYPE_UINT64 -DWITH_OPENSSL -D__TRANSMISSION__ -I/home/test/transmission/libtransmission/.. -I/home/test/transmission/build_with_devel_fmt/libtransmission/.. -isystem /home/test/transmission/third-party/fmt/include -isystem /home/test/transmission/third-party/small/include -isystem /home/test/transmission/build_with_devel_fmt/third-party/libevent.bld/pfx/include -isystem /home/test/transmission/third-party/googletest/googletest/include -isystem /home/test/transmission/build_with_devel_fmt/third-party/dht.bld/pfx/include -isystem /home/test/transmission/third-party/wide-integer -O2 -g -DNDEBUG -std=gnu++17 -W -Wall -Wextra -Wcast-align -Wextra-semi -Wextra-semi-stmt -Wextra-tokens -Wfloat-equal -Wgnu -Winit-self -Wint-in-bool-context -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wreorder-ctor -Wreturn-std-move -Wself-assign -Wself-move -Wsemicolon-before-method-body -Wsentinel -Wshadow -Wsign-compare -Wsometimes-uninitialized -Wstring-conversion -Wsuggest-destructor-override -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -Wno-sign-compare -o tests/libtransmission/CMakeFiles/libtransmission-test.dir/crypto-test.cc.o -c /home/test/transmission/tests/libtransmission/crypto-test.cc -v
    OpenBSD clang version 16.0.6
    Target: amd64-unknown-openbsd7.7
    Thread model: posix
    InstalledDir: /usr/bin
     (in-process)
     "/usr/bin/c++" -cc1 -triple amd64-unknown-openbsd7.7 -emit-obj -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name crypto-test.cc -mrelocation-model pic -pic-level 1 -pic-is-pie -mframe-pointer=all -relaxed-aliasing -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=constructor -dwarf-version=2 -debugger-tuning=gdb -v -fcoverage-compilation-dir=/home/test/transmission/build_with_devel_fmt -resource-dir /usr/lib/clang/16 -isystem /home/test/transmission/third-party/fmt/include -isystem /home/test/transmission/third-party/small/include -isystem /home/test/transmission/build_with_devel_fmt/third-party/libevent.bld/pfx/include -isystem /home/test/transmission/third-party/googletest/googletest/include -isystem /home/test/transmission/build_with_devel_fmt/third-party/dht.bld/pfx/include -isystem /home/test/transmission/third-party/wide-integer -D FMT_EXCEPTIONS=0 -D FMT_HEADER_ONLY=1 -D LIBTRANSMISSION_TEST_ASSETS_DIR=/home/test/transmission/tests/libtransmission/assets -D SMALL_DISABLE_EXCEPTIONS=1 -D WIDE_INTEGER_HAS_LIMB_TYPE_UINT64 -D WITH_OPENSSL -D __TRANSMISSION__ -I /home/test/transmission/libtransmission/.. -I /home/test/transmission/build_with_devel_fmt/libtransmission/.. -D NDEBUG -internal-isystem /usr/include/c++/v1 -internal-isystem /usr/lib/clang/16/include -internal-externc-isystem /usr/include -O2 -W -Wall -Wextra -Wcast-align -Wextra-semi -Wextra-semi-stmt -Wextra-tokens -Wfloat-equal -Wgnu -Winit-self -Wint-in-bool-context -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wreorder-ctor -Wreturn-std-move -Wself-assign -Wself-move -Wsemicolon-before-method-body -Wsentinel -Wshadow -Wsign-compare -Wsometimes-uninitialized -Wstring-conversion -Wsuggest-destructor-override -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wformat-security -Wno-sign-compare -std=gnu++17 -fdeprecated-macro -fdebug-compilation-dir=/home/test/transmission/build_with_devel_fmt -ferror-limit 19 -fwrapv -D_RET_PROTECTOR -ret-protector -fcf-protection=branch -fno-jump-tables -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -vectorize-loops -vectorize-slp -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-valloc -fno-builtin-free -fno-builtin-strdup -fno-builtin-strndup -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o tests/libtransmission/CMakeFiles/libtransmission-test.dir/crypto-test.cc.o -x c++ /home/test/transmission/tests/libtransmission/crypto-test.cc
    clang -cc1 version 16.0.6 based upon LLVM 16.0.6 default target amd64-unknown-openbsd7.7
    #include "..." search starts here:
    #include <...> search starts here:
     /home/test/transmission/libtransmission/..
     /home/test/transmission/build_with_devel_fmt/libtransmission/..
     /home/test/transmission/third-party/fmt/include
     /home/test/transmission/third-party/small/include
     /home/test/transmission/build_with_devel_fmt/third-party/libevent.bld/pfx/include
     /home/test/transmission/third-party/googletest/googletest/include
     /home/test/transmission/build_with_devel_fmt/third-party/dht.bld/pfx/include
     /home/test/transmission/third-party/wide-integer
     /usr/include/c++/v1
     /usr/lib/clang/16/include
     /usr/include
    End of search list.
    ```

    </details>

jggimi commented:
@tearfur Thank you for the extended testing.  I don't know how to isolate the problem further, either.

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
build: remove unused parameters from tr_add_external_auto_library

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
build: update crc32c cmake script

rafe-s commented:
As drafted, this breaks things on FreeBSD in ways that I have not yet been able to resolve. 

CMAKE gets confused and expects FindDHT.cmake to be named Finddht.cmake. This can be fixed by revising `tr_add_external_auto_library(DHT dht` to `tr_add_external_auto_library(DHT DHT dht` in CMAKELists.txt.

What I have not been able to resolve, however, are issues it has with libutp. For some reason, this merge results in CMAKE reporting, 

```
CMake Error at libtransmission/CMakeLists.txt:271 (target_link_libraries):
  Target "transmission" links to:
    libutp::libutp
  but the target was not found.  Possible reasons include:
    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

This occurs after CMAKE reported earlier during its configuration that it found libutp at /usr/local/lib/libutp.so, which is Transmission's current fork of libutp.

tearfur committed:
fix: dht system library

tearfur committed:
fix: add libutp interface target

tearfur committed:
Merge branch 'main' into third-party

rafe-s commented:
@tearfur Yes! This builds cleanly and passes all tests in a clean build environment. This is true both when I build it as a port and when I build it manually using devel/dht. This is great news, as we have been patching CMake quite regularly to get Transmission to build with system dependencies on Freebsd, and this PR would fix that. I will note that my build system tracks FreeBSD 14-STABLE, not 14.3-RELEASE, but there is zero reason why that should make a difference for the dependencies involved as they are all ports. Build log here: [transmission-daemon-4.1.b3_3.log](https://github.com/user-attachments/files/23288936/transmission-daemon-4.1.b3_3.log)

I was unable to reproduce your the error in my build environment. Is it possible that this a googletest issue, not a dht issue?

rafe-s commented:
Looking at your build log, it looks like this is before you implemented all your patches? The Finddht module is not properly named?

```

2025-10-31T13:32:58.8289584Z CMake Warning (dev) at /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:441 (message):
2025-10-31T13:32:58.8290631Z   The package name passed to `find_package_handle_standard_args` (DHT) does
2025-10-31T13:32:58.8291393Z   not match the name of the calling package (dht).  This can lead to problems
2025-10-31T13:32:58.8292117Z   in calling code that expects `find_package` result variables (e.g.,
2025-10-31T13:32:58.8292684Z   `_FOUND`) to follow a certain pattern.
2025-10-31T13:32:58.8293124Z Call Stack (most recent call first):
2025-10-31T13:32:58.8293653Z   cmake/Finddht.cmake:27 (find_package_handle_standard_args)
2025-10-31T13:32:58.8294221Z   cmake/TrMacros.cmake:140 (find_package)
2025-10-31T13:32:58.8294731Z   CMakeLists.txt:553 (tr_add_external_auto_library)
2025-10-31T13:32:58.8295361Z This warning is for project developers.  Use -Wno-dev to suppress it.
2025-10-31T13:32:58.8295784Z 
2025-10-31T13:32:58.8297022Z -- Found DHT: /usr/local/lib/libdht.so

```

rafe-s commented:
@mikedld It would be great to have this land in the next beta as it would ease BSD port administration considerably.

ckerr commented:
@tearfur could you  update the PR to fix the merge conflict in `cmake/FindFmt.cmake`?

@mikedld last chance to request any changes :smile_cat:

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
Merge branch 'main' into third-party

mikedld commented on cmake/TrMacros.cmake:
```diff
@@ -128,48 +128,95 @@ function(tr_process_list_conditions VAR_PREFIX)
 
-macro(tr_add_external_auto_library ID DIRNAME LIBNAME)
-    cmake_parse_arguments(_TAEAL_ARG "SUBPROJECT" "TARGET" "CMAKE_ARGS" ${ARGN})
+macro(tr_get_fmt_version OVAR INCLUDE_DIR)
```
I'd move this logic to cmake/Findfmt.cmake — making find module depend on TrMacros.cmake doesn't feel right given that such modules should generally be reusable (even if don't expect them to be) — and define both `FMT_USE_EXCEPTIONS` and `FMT_EXCEPTIONS` to simplify things. We know which version we bundle so parsing headers shouldn't be necessary in that case.

tearfur replied:
8ea0cbedadfca368a81aa03553dec5ce9d439afc

tearfur committed:
code review: move tr_get_fmt_version into Findfmt.cmake

tearfur committed:
code review: use option() for gtest

tearfur committed:
code review: move TrGtest.cmake

tearfur committed:
code review: move find_package(PkgConfig) out of loop

tearfur committed:
build: delete FindCrc32c.cmake

Impossible to parse package version from distributed source files.

tearfur committed:
code review: Finddht.cmake

tearfur committed:
build: delete FindFastFloat.cmake

Impossible to parse package version from distributed source files.

tearfur committed:
code review: Findfmt.cmake

tearfur committed:
code review: Findlibb64.cmake

tearfur committed:
code review: Findlibdeflate.cmake

tearfur committed:
code review: Findlibevent.cmake

tearfur committed:
code review: Findlibnatpmp.cmake

tearfur committed:
code review: Findlibpsl.cmake

tearfur committed:
code review: Findlibutp.cmake

tearfur committed:
code review: Findlibminiupnpc.cmake

tearfur committed:
code review: FindRapidJSON.cmake

tearfur committed:
build: delete FindSmall.cmake

Impossible to parse package version from distributed source files.

tearfur committed:
build: only accept cmake config package for utf8cpp

Impossible to parse package version from distributed source files.

tearfur committed:
build: delete FindWideInteger.cmake

Impossible to parse package version from distributed source files.

nevack commented:
Maybe I'm late to chime in with this, but:

There are now like 15 `USE_SYSTEM_*` options.
Is it possible to introduce one new option ~~to rule them all~~ `USE_SYSTEM`/`PREFER_SYSTEM` that controls global default of all `USE_SYSTEM_*` options?

`OFF` - do not use system deps, build everything from source and link statically.
`ON` - use everything from system and fail configuration if system library is not found.
`AUTO` - default value, current behaviour, to use or not to use system is controlled per dep.

When I need to test build with deps built from source, it's tiresome to toggle off 15 options.
The same is true for github actions, where we manually specify OFF for all deps.

@tearfur @mikedld WDYT?

tearfur commented on cmake/TrGTest.cmake:
953a126d53d0438494f8a3ee949e4e2ec5d5edd5

tearfur committed:
build: add `USE_SYSTEM_DEFAULT`

tearfur committed:
Merge branch 'main' into third-party

ckerr commented:
The has been in `needs code review` limbo for awhile now and has had contributor testing & feedback.

Unless @mikedld or anyone else has more review feedback, I'll merge this in a couple of days.

ckerr committed:
Merge branch 'main' into third-party

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
ci: drop Fedora 40 and adopt Fedora 43

tearfur committed:
ci: try to silence system header warnings

ckerr commented on .github/workflows/actions.yml:
```diff
@@ -454,6 +454,8 @@ jobs:
             -DCMAKE_BUILD_TYPE=RelWithDebInfo \
             -DCMAKE_INSTALL_PREFIX=pfx \
             -DCMAKE_PREFIX_PATH=`brew --prefix`/opt/qt \
+            -DCMAKE_CXX_FLAGS='--system-header-prefix=/usr/local/include' `# Needed to supress system header warnings in macos-15-intel` \
```
I _think_ this should be `--system-header-prefix=fmt/` rather than the absolute path?

tearfur replied:
Nice. That worked.

Look at the warnings again, I really don't think that's an error on our part. All we did was taking the target from `fmt-config.cmake` then linked to it, so those headers file should have been labelled as system headers without `--system-header-prefix` or other workarounds. All other runners were fine too.

tearfur committed:
ci: use `cmake --install`

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
Merge branch 'main' into third-party

rafe-s commented:
Commit [2158d63](https://github.com/transmission/transmission/pull/7554/commits/2158d631fd90c4a18c3290b7c08c3fc088345ed8) breaks the build for me. 

> CMake Error at /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
>   Could NOT find utf8cpp (missing: utf8cpp_INCLUDE_DIR)
> Call Stack (most recent call first):
>   /usr/local/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
>   cmake/Findutf8cpp.cmake:16 (find_package_handle_standard_args)
>   cmake/TrMacros.cmake:154 (find_package)
>   CMakeLists.txt:642 (tr_add_external_auto_library)

rafe-s commented:
Same with commit 448c022 and wide-integer.

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
Revert "build: only accept cmake config package for utf8cpp"

This reverts commit 2158d631fd90c4a18c3290b7c08c3fc088345ed8.

tearfur committed:
build: harden utf8cpp find module

tearfur commented:
@rafe-s Not `CMakeLists.txt`, but `utf8cppConfig.cmake`.

* * *

For utf8cpp, there are already released ports that does not have `utf8cppConfig.cmake`, so I have no choice but to keep 24592d8876bf7cb6ebfeb9e2bdf51ca2255da1e2.

But for wide-integer, there are no existing OS packages for it in all the systems we test on. If the first released port on FreeBSD will include `WideIntegerConfig.cmake`, then I prefer to not have `FindWideInteger.cmake` at all.

tearfur commented:
The FreeBSD and OpenBSD issues should be fixed now.

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
chore: bump wide-integer

Pick up https://github.com/ckormanyos/wide-integer/commit/bf9398f9dadf60aa3928be6a0681c2d5ebdad457 and https://github.com/ckormanyos/wide-integer/commit/bcc726a30f3d7c475266b34e83ff50aa9186f285

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
refactor: gtest should be included with angled brackets

Now that gtest is built as a system library, it should be included with angled brackets instead of quotes.

tearfur committed:
Merge branch 'main' into third-party

rafe-s commented:
I can confirm this builds clean on FreeBSD as well.

tearfur committed:
Merge branch 'main' into third-party

ckerr commented:
> @ckerr What's the status on this? Are we still waiting for @mikedld's re-review?

Yes.

This is a very large change to a part of the code that @mikedld is the primary on, so I'd rather defer to him on this instead of going around him, especially since he did review this PR last month.

So, @mikedld ping :smile_cat:

tearfur committed:
Merge branch 'main' into third-party

tearfur committed:
code review: fixup libutp variables before `find_package_handle_standard_args`

tearfur committed:
code review: define `WIDE_INTEGER_HAS_LIMB_TYPE_UINT64` only for targets depending on WideInteger

tearfur committed:
chore: bump wide-integer

Pickup https://github.com/ckormanyos/wide-integer/commit/4b2258acac771baf3e9961be6bb89c9a907a2634 so that wide-integer tests won't run in Transmission project.

tearfur committed:
Merge branch 'main' into third-party

# Conflicts:
#	tests/libtransmission/announcer-test.cc
#	tests/libtransmission/benc-test.cc
#	tests/libtransmission/block-info-test.cc
#	tests/libtransmission/buffer-test.cc
#	tests/libtransmission/completion-test.cc
#	tests/libtransmission/file-piece-map-test.cc
#	tests/libtransmission/getopt-test.cc
#	tests/libtransmission/ip-cache-test.cc
#	tests/libtransmission/magnet-metainfo-test.cc
#	tests/libtransmission/net-test.cc
#	tests/libtransmission/peer-mgr-wishlist-test.cc
#	tests/libtransmission/quark-test.cc
#	tests/libtransmission/session-alt-speeds-test.cc
#	tests/libtransmission/settings-test.cc
#	tests/libtransmission/strbuf-test.cc
#	tests/libtransmission/tr-peer-info-test.cc
#	tests/libtransmission/values-test.cc
#	tests/libtransmission/web-utils-test.cc

tearfur committed:
Merge branch 'main' into third-party

ckerr commented:
Oop I missed that this was approved. Sorry about that.
